README.md
# react-native-argon2

React Native Wrapper around native Argon2 implementations:

iOS: [CatCrypto](//github.com/ImKcat/CatCrypto)

Android: [argon2kt](//github.com/lambdapioneer/argon2kt)

## Getting started
```bash
npm install @sphereon/react-native-argon2 --save
```

Compatibility Table

| React Native Version | Package Version |
|----------------------|-----------------|
| 0.60 - 0.63.2        | ~0.1.0         |
| 0.63.3               | ^1.0.0         |

## Usage

```javascript
import argon2 from '@sphereon/react-native-argon2';
const password = 'password';
const salt = '1234567891011121314151617181920212223242526272829303132333435363';
const result = await argon2(password, salt, {});
const { rawHash, encodedHash } = result;
// rawHash: 031d6c82ddede1200f4794605052745dd562bd4db358e23dac1b11c052eff8d9
// encodedHash: $argon2id$v=19$m=32768,t=2,p=1$MTIzNDU2Nzg5MTAxMTEyMTMxNDE1MTYxNzE4MTkyMDIxMjIyMzI0MjUyNjI3MjgyOTMwMzEzMjMzMzQzNTM2Mw$Ax1sgt3t4SAPR5RgUFJ0XdVivU2zWOI9rBsRwFLv+Nk
```

### Input
The package takes in the following variables:

| Parameter           | Type    |
|---------------------|---------|
| password            | string  |
| salt                | string  |
| config              | object  |
| config.iterations   | integer |
| config.memory       | integer |
| config.parallelism  | integer |
| config.hashLength   | integer |
| config.mode         | string  |

You are not required to configure the third parameter which is the `config` object, however you do have to provide an empty object to it if you are not changing any of the values. You can set config values with the following example:

```javascript
const result = await argon2(
    password,
    salt,
    {
      iterations: 5,
      memory: 16 * 1024,
      parallelism: 2,
      hashLength: 20,
      mode: 'argon2i'
    }
);
```


### Output

`rawHash` is the hexadecimal representation

`encodedHash` is the string representation


CHANGELOG.md
# Changelog

### [2.0.9](https://github.com/poowf/react-native-argon2/compare/v2.0.7...v2.0.9) (2023-03-11)


### Maintenance

* prep for next dev cycle ([92a01a8](https://github.com/poowf/react-native-argon2/commit/92a01a89d0498dd96ad97b0131fa46794618577b))
* update readme and yarn.lock ([93e2a7a](https://github.com/poowf/react-native-argon2/commit/93e2a7ad53cf2da1502ac1a2dc1b633e337961cc))

### [2.0.7](https://github.com/poowf/react-native-argon2/compare/v2.0.1...v2.0.7) (2022-08-23)


### Maintenance

* update yarn.lock ([9345aaa](https://github.com/poowf/react-native-argon2/commit/9345aaa447263e07a0afa589ab88a7b6b9bb561c))
* use maven central instead of jcenter ([269c64e](https://github.com/poowf/react-native-argon2/commit/269c64e91b188222ad60790708d4d9c36a3852a6))

### [2.0.1](https://github.com/poowf/react-native-argon2/compare/v2.0.0...v2.0.1) (2021-06-20)


### Maintenance

* bump packages ([0458d07](https://github.com/poowf/react-native-argon2/commit/0458d07ae5059d806003fda685554fa7193548a6))

## [2.0.0](https://github.com/poowf/react-native-argon2/compare/v1.1.1...v2.0.0) (2020-10-24)


### ⚠ BREAKING CHANGES

* Switch to argon2kt for Android due to licensing issues
* Switch to CatCrypto for iOS due to licensing issues

### Features

* Implement configurable hasher ([099a8d5](https://github.com/poowf/react-native-argon2/commit/099a8d564c74ed5576da177b96dd54eb2b1e3c01))


* Switch to argon2kt for Android due to licensing issues ([d3e575b](https://github.com/poowf/react-native-argon2/commit/d3e575b8bc15787a14485533b132820ddeac608d))
* Switch to CatCrypto for iOS due to licensing issues ([0ad6598](https://github.com/poowf/react-native-argon2/commit/0ad65981891794c5ade69633e2d1eaba0714d5c5))

### [1.1.1](https://github.com/poowf/react-native-argon2/compare/v1.1.0...v1.1.1) (2020-10-22)


### Maintenance

* Update README ([df047a4](https://github.com/poowf/react-native-argon2/commit/df047a43377b5b3dfeac787a163956a45d0d5302))

## 1.1.0 (2020-10-22)


### Features

* Add standard-version for releases ([f057b8e](https://github.com/poowf/react-native-argon2/commit/f057b8efa2e64ce7b4c429959cbf532c3b13ebd1))
* Use published cocoapods instead ([a8b4101](https://github.com/poowf/react-native-argon2/commit/a8b410116c803660f85d2a7d2f223a217856e3d9))


### Bug Fixes

* Failing builds on XCode 12 ([19eeaee](https://github.com/poowf/react-native-argon2/commit/19eeaee6b3e1a46f9c39bb8fd29455d8113cc392))
* XCode project warnings ([f40bc7b](https://github.com/poowf/react-native-argon2/commit/f40bc7bcc5a38141048304b0f657da3a9bd04c71))


### Maintenance

* Update README ([ae3f183](https://github.com/poowf/react-native-argon2/commit/ae3f183d1cf92c7b451859efd3206fd21a9a3d30))
* **release:** 1.0.0 ([75b61c4](https://github.com/poowf/react-native-argon2/commit/75b61c4a4c1ddc6085a0482258ff3cd60be7bfc6))
* Bump version ([e559de2](https://github.com/poowf/react-native-argon2/commit/e559de2358302902f7b7574ddef9668a63d8ae90))

### 1.0.0 (2020-10-18)


### Features

* Add standard-version for releases ([f057b8e](https://github.com/poowf/react-native-argon2/commit/f057b8efa2e64ce7b4c429959cbf532c3b13ebd1))
* Use published cocoapods instead ([a8b4101](https://github.com/poowf/react-native-argon2/commit/a8b410116c803660f85d2a7d2f223a217856e3d9))


### Bug Fixes

* Failing builds on XCode 12 ([19eeaee](https://github.com/poowf/react-native-argon2/commit/19eeaee6b3e1a46f9c39bb8fd29455d8113cc392))
* XCode project warnings ([f40bc7b](https://github.com/poowf/react-native-argon2/commit/f40bc7bcc5a38141048304b0f657da3a9bd04c71))


ios/RNArgon2.swift
import Foundation
import CatCrypto

@objc(RNArgon2)
class RNArgon2: NSObject {
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc
  func argon2(_ password: String, salt: String, config: NSDictionary? = nil, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
    let argon2Context = CatArgon2Context.init();
    let configDict = config as! Dictionary<String,Any>

    argon2Context.iterations = configDict["iterations", default: 2 ] as! Int;
    argon2Context.memory = configDict["memory", default: 32 * 1024 ] as! Int;
    argon2Context.parallelism = configDict["parallelism", default: 1 ] as! Int;
    argon2Context.salt = salt.hexaBytes();
    argon2Context.hashLength = configDict["hashLength", default: 32 ] as! Int;
    argon2Context.mode = getArgon2Mode(mode: configDict["mode", default: "argon2id" ] as! String);

    let argon2Crypto = CatArgon2Crypto.init(context: argon2Context);
    let encodedResult = argon2Crypto.hash(password: password);

    // Set the hasher to hash as raw since encoded is the default
    argon2Crypto.context.hashResultType = .hashRaw;
    let rawResult = argon2Crypto.hash(password: password);

    if ((rawResult.error) != nil || (encodedResult.error) != nil) {
        let error = NSError(domain: "com.poowf.argon2", code: 200, userInfo: ["Error reason": "Failed to generate argon2 hash"])
        reject("E_ARGON2", "Failed to generate argon2 hash", error)
    }

    let rawHash = rawResult.hexStringValue();
    let encodedHash = encodedResult.stringValue();

    let resultDictionary: NSDictionary = [
        "rawHash" : rawHash,
        "encodedHash" : encodedHash,
    ]
    resolve(resultDictionary);
  }

  func getArgon2Mode(mode: String) -> CatArgon2Mode {
    var selectedMode: CatArgon2Mode;
    switch mode {
        case "argon2d":
            selectedMode = CatArgon2Mode.argon2d;
            break;
        case "argon2i":
            selectedMode = CatArgon2Mode.argon2i;
            break;
        case "argon2id":
            selectedMode = CatArgon2Mode.argon2id;
            break;
        default:
            selectedMode = CatArgon2Mode.argon2id;
            break;
    }

    return selectedMode;
  }
}


android/src/main/java/com/poowf/argon2/RNArgon2Module.java
package com.poowf.argon2;

import android.app.Activity;
import android.content.Intent;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.facebook.react.bridge.Promise;

import com.lambdapioneer.argon2kt.Argon2Kt;
import com.lambdapioneer.argon2kt.Argon2KtResult;
import com.lambdapioneer.argon2kt.Argon2Mode;

import java.math.BigInteger;

public class RNArgon2Module extends ReactContextBaseJavaModule {
    private ReactContext mReactContext;

    public RNArgon2Module(ReactApplicationContext reactContext) {
        super(reactContext);
        mReactContext = reactContext;
    }

    @Override
    public String getName() {
        return "RNArgon2";
    }

    @ReactMethod
    public void argon2(String password, String salt, ReadableMap config, Promise promise) {
        try {
            final byte[] passwordBytes = password.getBytes("UTF-8");

            final byte[] saltInputBytes = new BigInteger(salt, 16).toByteArray();
            byte[] saltBytes = new byte[32];
            System.arraycopy(saltInputBytes, saltInputBytes.length - 32, saltBytes, 0, 32);

            Integer iterations = config.hasKey("iterations") ? new Integer(config.getInt("iterations")) : new Integer(2);
            Integer memory = config.hasKey("memory") ? new Integer(config.getInt("memory")) : new Integer(32 * 1024);
            Integer parallelism = config.hasKey("parallelism") ? new Integer(config.getInt("parallelism")) : new Integer(1);
            Integer hashLength = config.hasKey("hashLength") ? new Integer(config.getInt("hashLength")) : new Integer(32);
            Argon2Mode mode = config.hasKey("mode") ? getArgon2Mode(config.getString("mode")) : Argon2Mode.ARGON2_ID;

            final Argon2Kt argon2Kt = new Argon2Kt();

            final Argon2KtResult hashResult = argon2Kt.hash(
                    mode,
                    passwordBytes,
                    saltBytes,
                    iterations,
                    memory,
                    parallelism,
                    hashLength);
            final String rawHash = hashResult.rawHashAsHexadecimal(false);
            final String encodedHash = hashResult.encodedOutputAsString();

            WritableMap resultMap = new WritableNativeMap();
            resultMap.putString("rawHash", rawHash);
            resultMap.putString("encodedHash", encodedHash);

            promise.resolve(resultMap);
        } catch (Exception exception) {
            promise.reject("Failed to generate argon2 hash", exception);
        }
    }

    public Argon2Mode getArgon2Mode(String mode) {
        Argon2Mode selectedMode;
        switch (mode) {
            case "argon2d":
                selectedMode = Argon2Mode.ARGON2_D;
                break;
            case "argon2i":
                selectedMode = Argon2Mode.ARGON2_I;
                break;
            case "argon2id":
                selectedMode = Argon2Mode.ARGON2_ID;
                break;
            default:
                selectedMode = Argon2Mode.ARGON2_ID;
                break;
        }

        return selectedMode;
    }
}
